rust/src/types/observer.rs

```diff
@@ -0,0 +1,52 @@
+use std::borrow::Cow;
+
+use crate::interpreter::Interpreter;
+
+pub trait Observer {
+    fn pre_op(&mut self, interpreter: &Interpreter);
+
+    fn post_op(&mut self, interpreter: &Interpreter);
+
+    fn log(&mut self, message: Cow<str>);
+}
+
+pub struct NoOpObserver();
+
+impl Observer for NoOpObserver {
+    fn pre_op(&mut self, _interpreter: &Interpreter) {}
+
+    fn post_op(&mut self, _interpreter: &Interpreter) {}
+
+    fn log(&mut self, _message: Cow<str>) {}
+}
+
+pub struct LoggingObserver();
+
+impl Observer for LoggingObserver {
+    fn pre_op(&mut self, interpreter: &Interpreter) {
+        println!(
```

Comment:
Please, use a file descriptor, which we can set to stdout or a file sync. 
These are the standard options implemented in lfvm.